### PR TITLE
[IMP] stock: reserve on return to select delivered lot

### PR DESCRIPTION
**Pull request by amoyaux**
Since return moves are linked to their original move, it's easy to
manage reservation in those case. We just don't bypass the reservation
when it comes from a reservation outside our stock in order to manage
the computation on linked moves and reserve pieces that were not yet
return.

Also add a return picking type for 2 reasons:
- Allow to select existing lot.
- Display reserved move line for an incoming picking.

Note that this picking type is just a default configuration and could be
modify by the user without any issue.

### CODE REVIEW

File: addons/stock/models/stock_move_line.py [hunk 1]
```diff
@@ -693,7 +693,7 @@ def _free_reservation(self, product_id, location_id, quantity, lot_id=None, pack
 
     def _should_bypass_reservation(self, location):
         self.ensure_one()
-        return location.should_bypass_reservation() or self.product_id.type != 'product'
+        return self.move_id._should_bypass_reservation() or self.product_id.type != 'product'
```
Comment (svs-odoo) [inline addons/stock/models/stock_move_line.py:696]: You could remove the `self.product_id.type != 'product'` part as you already check it in the `stock.move` `_should_bypass_reservation`, no ?
```suggestion
        return self.move_id._should_bypass_reservation()
```

File: addons/stock/models/stock_move.py [hunk 2]
```diff
@@ -1335,10 +1377,26 @@ def _action_assign(self):
...
-                if move.product_id.tracking == 'serial' and (move.picking_type_id.use_create_lots or move.picking_type_id.use_existing_lots):
+                if move.move_orig_ids:
+                    available_move_lines = _get_available_move_lines(move)
+                    for (location_id, lot_id, package_id, owner_id), quantity in available_move_lines.items():
+                        qty_added = min(missing_reserved_quantity, quantity)
+                        move_line_vals = move._prepare_move_line_vals(qty_added)
+                        move_line_vals.update({
+                            'location_id': location_id.id,
+                            'lot_id': lot_id.id,
+                            'lot_name': lot_id.name,
```
Comment (Whenrow) [inline addons/stock/models/stock_move.py:1388]: missing package too ?

File: addons/stock/models/stock_move.py [hunk 3]
```diff
@@ -1335,10 +1377,26 @@ def _action_assign(self):
...
+                            'owner_id': owner_id.id,
+                        })
+                        move_line_vals_list.append(move_line_vals)
+                        missing_reserved_quantity -= qty_added
+                        if float_is_zero(missing_reserved_quantity, precision_rounding=move.product_uom.rounding):
```
Comment (Whenrow) [inline addons/stock/models/stock_move.py:1393]: I would use the move line uom instead

File: addons/stock/tests/test_packing.py [hunk 4]
```diff
@@ -218,7 +218,7 @@ def test_put_in_pack_to_different_location(self):
         })
         picking = self.env['stock.picking'].create({
             'picking_type_id': self.warehouse.in_type_id.id,
-            'location_id': self.stock_location.id,
+            'location_id': self.customer_location.id,
```
Comment (Whenrow) [inline addons/stock/tests/test_packing.py:221]: Manque un pti test non ?

File: addons/stock/models/stock_move.py (hunk 2 above)
Comment (amoyaux) [reply addons/stock/models/stock_move.py:1388]: It does not really make sense to copy the package, the customer probably unpack it and don't return the whole order

File: addons/stock/models/stock_move_line.py [hunk 5]
```diff
@@ -351,7 +351,7 @@ def write(self, vals):
                             raise
 
                 # Reserve the maximum available of the new charateristics of the move line.
-                if not ml._should_bypass_reservation(updates.get('location_id', ml.location_id)):
+                if not ml.move_id._should_bypass_reservation():
```
Comment (amoyaux) [inline addons/stock/models/stock_move_line.py:354]: Foireux ça

File: addons/stock/models/stock_move.py [hunk 6]
```diff
@@ -1349,10 +1391,26 @@ def _action_assign(self):
...
+                        move_line_vals = move._prepare_move_line_vals(qty_added)
+                        move_line_vals.update({
+                            'location_id': location_id.id,
+                            'lot_id': lot_id.id,
+                            'lot_name': lot_id.name,
+                            'owner_id': owner_id.id,
+                        })
+                        move_line_vals_list.append(move_line_vals)
+                        missing_reserved_quantity -= qty_added
+                        if float_is_zero(missing_reserved_quantity, precision_rounding=move.product_id.uom_id.rounding):
```
Comment (Whenrow) [inline addons/stock/models/stock_move.py:1407]: Maybe more secure to do a `float_compare(missing_reserve_quantity, 0) <= 0`

File: addons/stock/models/stock_move.py [hunk 7]
```diff
@@ -1349,10 +1391,26 @@ def _action_assign(self):
...
+                            break
+
+                if missing_reserved_quantity and move.product_id.tracking == 'serial' and (move.picking_type_id.use_create_lots or move.picking_type_id.use_existing_lots):
                     for i in range(0, int(missing_reserved_quantity)):
                         move_line_vals_list.append(move._prepare_move_line_vals(quantity=1))
-                else:
+                elif missing_reserved_quantity:
```
Comment (Whenrow) [inline addons/stock/models/stock_move.py:1413]: You test missing_reserved_quantity but you add missing_reserved_**uom**_quantity below

**Review by Whenrow — APPROVED** [review]

**Comment by amoyaux** [comment]
robodoo r+

**Comment by mart-e** [comment]
I think this breaks an enterprise test https://runbot.odoo.com/runbot/build/9791948

**Comment by xmo-odoo** [comment]
robodoo r-

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
https://github.com/odoo/enterprise/pull/20605

**Comment by pimodoo** [comment]
@robodoo retry

**Comment by d-fence** [comment]
robodoo retry

**Pull request closed without merging**